### ISSUE TITLE
drop the https:// or git:// protocol definition in the repo section and automatically detect

### ISSUE THREAD

**Issue by Raffaello** [pre-commit/pre-commit#1867]
I think it would be possible to crete a small algorithm to check if the url provided is not starting with any https:// or git:// protocol specify and then it will autodetect which one to use starting from the git:// one and if no ssh configured can fallback to using https mechanism.

Otherwise in this way it is forced to use a repo only with git or https and based on different users one or the other could be prefered.

basically could be added at the time of cloning the remote repo to do this check to determine the protocol to use if not specified.

**Comment by asottile** [comment]
please describe what you're trying to solve -- what's the problem with specifying `https` ?

**Comment by asottile** [comment]
but no, pre-commit will not be parsing and changing git urls -- that sounds like a debugging nightmare -- it will clone literally what you ask it to clone

perhaps you're interested in git's `insteadOf` configuration option? https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlltbasegtinsteadOf

**Comment by Raffaello** [comment]
> but no, pre-commit will not be parsing and changing git urls -- that sounds like a debugging nightmare -- it will clone literally what you ask it to clone
> 
> perhaps you're interested in git's `insteadOf` configuration option? https://git-scm.com/docs/git-config#Documentation/git-config.txt-urlltbasegtinsteadOf

i can understand that, but it was simpler than chainging git url, it was more com[leting them like:
eg:
```
repos:
-   repo: https://github.com/pre-commit/pre-commit-hooks
```
instead when:
```
repos:
-   repo: github.com/pre-commit/pre-commit-hooks
```

it will complete the starting url with either `git://` and `https://`


but yes it could be a nightmare, i can understand that....

----

Yes the `insteadOf` is that use case, would be possible to implement in pre-commit with that if it is failing to clone? 
maybe in the yaml file specifing some fallbacks or some sort of those? 

otherwise the insteadOf must be applied globally as the repo of pre-commit that is referring to, might not yet be cloned and can't set up just for that.

**Comment by asottile** [comment]
`git://` is never the right answer, since it is insecure

the thing in `repo` can be any sort of string, it might be a path, it might be on github, it might be on some server -- we can't realistically change the string there and expect it to work (or even expect it to point to the same thing!)

**Comment by Raffaello** [comment]
> `git://` is never the right answer, since it is insecure
> 
> the thing in `repo` can be any sort of string, it might be a path, it might be on github, it might be on some server -- we can't realistically change the string there and expect it to work (or even expect it to point to the same thing!)

understand that,  anyway i was guessing as this repo should have at least a way to support both the github way of cloning as a clone is required to initialize hooks, in this case is need to have access to a 3rd party repo, if it is private require some authentication to retrieve it, that where the problem come from as it asks an authentication to initialize the 4rd party repo during a `git commit` not really nice, as the initialization should be done when installing the hooks, rather than lazy initialization at the first time is used. anyway.

Hopefully GitHub will drop the git:// protocol then! 

thanks.

**Comment by asottile** [comment]
> as the initialization should be done when installing the hooks, rather than lazy initialization at the first time is used. anyway.

there's an option for this, it's `--install-hooks` -- but it's not possible to do what you're saying "should" be done because the `.pre-commit-config.yaml` file could be altered after you run `pre-commit install`.

> if it is private require some authentication to retrieve it, that where the problem come from

if it's a private repo your configuration should always use an ssh url, if it's a public repo your configuration should always use https://
